### ISSUE TITLE
1.7 crashes after entering password ios10.3.3 iphone 6s

### ISSUE THREAD

**Issue by dbush2017** [MiniKeePass/MiniKeePass#526]
Updated today to 1.7 on my iPhone 6s running iOS 10.3.3. I can connect to datebase but cannot retrieve any entries.

**Comment by starfotograf** [comment]
I had the same problem, since update to 1.7 the app crashes after entering password. Solutions?

**Comment by starfotograf** [comment]
Exporting the KDB to v1.x on the PC with KeePass 2.36 will open on the iPhone iOS, but I did not find a search function. Any other better solution?

**Comment by jamz216** [comment]
I am having the same problem with both my iPhone and iPad

**Comment by janus-matthew** [comment]
It crashes on iOS 11.0.1 too. I updated the OS today in the hope that since the latest app update was to support iOS 11, that might fix things.  Unfortunately not...

**Comment by jarush** [comment]
Since the couple password databases I tested don’t crash on my device, could someone that’s reported this issue send me a sample database that crashes? Just email me a simple kdb file with no personal information, along with the password and I’ll be able to debug the issue. You can send it to jarush <at> gmail <dot> com. Thanks.

**Comment by gabox7** [comment]
Im having same problem with Ver1.7 & 6s plus (10.3.3). Please any workaround ? thanks in advance.

**Comment by dbush2017** [comment]
All of my kdb files have personal information and I am not comfortable sharing. I can tell you that only 2 of the groups in my kdb file crash on my iphone6s. Both of them are large. The 2 groups that work contain a couple of entries only. and a third group is large that opens. The other thing I noticed is the ones that crash have subgroups. The ones that work do not. Hope this helps.

**Comment by Erreu** [comment]
Hi @jarush It might be related to an entry with a blank title? https://github.com/MiniKeePass/MiniKeePass/issues/537#issuecomment-333430017

**Comment by jarush** [comment]
Hmm.  I just tried creating a kdb 1.x and a kdbx 2.x files on my laptop, each with a single entry with no title using the KeePass desktop app.  They both opened up just fine in the simulator.  I was able to see the entry in the group, select the entry and view it's details, all without a title.

**Comment by Erreu** [comment]
@jarush I'll see if I can re-create a crashing database for you.

**Comment by tssmith7** [comment]
I've tried to reproduce a crashing database as well without success.  I've tried very large databases with 10 deep nested groups each with 1000 entries and a single group with 30,000 entries.  I've tried the empty Title field, custom rich text attachment, image file attachment, using a key file or not. etc.  Getting an example database that crashes is very very helpful....

**Comment by Erreu** [comment]
@tssmith7 Well, I know that I fixed my crashes by giving a title to any entries that didn't have a title, but jarush was right, just creating a "blank title" entry from scratch doesn't cause the issue. I might try to anonymize a stripped down backup of my real database that crashes if I can't make any headway on creating one from scratch.

**Comment by Erreu** [comment]
@jarush @tssmith7 Attached is a kdbx which reliably crashes my MiniKeePass 1.7.1 on IOS 10.3.3. The password is "password". I've also attached the KeePass XML 2.x file that contains the two records which I imported to a new kdbx cause the crash. My next step was going to be to narrow down the specific fields, but I ran out of time this evening. I created the xml by exporting and anonymizing two records from a backup that was causing crashes. Hopefully it crashes for you. 

[crash10.kdbx.zip](https://github.com/MiniKeePass/MiniKeePass/files/1347960/crash10.kdbx.zip)
[crash10.xml.zip](https://github.com/MiniKeePass/MiniKeePass/files/1347961/crash10.xml.zip)

**Comment by jarush** [comment]
@Erreu - Thanks! That sample file does crash it.  It does look like there are 2 entries that have no title which causes the issue.  Looking at the parsed XML after the file is decrypted, some of the elements are completely missing (e.g. the title for example).  That's interesting, because when I created a sample database with no title, it still had the title element but with an empty string for the title.  This should be fairly easy to fix.

**Comment by tssmith7** [comment]
@jarush, I found that if I converted the crash10.kdbx file to a kdb(1.x) file with MacPass that the resulting database crashes MKP.  However, if I use KeePass 2.36 (on Windows) to export a kdb(1x) file, the file does not crash MKP.  The behavior of KeePass 2.36 seems to be to leave any missing standard string fields alone.  With exporting a kdb(1.x) file from KeePass 2.36 all missing string fields are filled in as empty strings.  When I open the crash10.kdb file that causes MKP to crash with KeePass 1.32 and save as another name, the resulting file also does not crash MKP.  The cause of the crash is a NULL group name causing an exception when sorted (same as the kdbx(2.x) crash)

So, I don't know if you want to address the somewhat esoteric case of a kdb(1.x) created from a non-KeePass application that is missing string field definitions.  It seems like maybe it would be a good idea to give both the group.name field and the entry.title field defaults for kdb(1.x) files as well.

**Comment by tssmith7** [comment]
BTW @Erreu fantastic job with the crashing database.

**Comment by jarush** [comment]
@tssmith7 good catch, even if it’s an unlikely case, it’s safer if we just handle it. Could you attach the 1.x database that causes the crash?

**Comment by tssmith7** [comment]
Here is one for the blank group names.  I cannot seem to create the case with empty Entry Titles...
The password is 'test'

[twoblankgroups.kdb.zip](https://github.com/MiniKeePass/MiniKeePass/files/1351091/twoblankgroups.kdb.zip)

**Comment by Erreu** [comment]
Thanks @tssmith7, it's nothing really,  I'm naturally talented at breaking things. :smiling_imp:

**Comment by tssmith7** [comment]
@jarush Would using an "optional chaining" operator on the sorting call fix all cases?  I'm not that familiar with Swift and I don't understand the trade-offs.  Is this what Objective-C does by default?  If somewhere along the chain is NULL it just returns NULL (nil).  I'm a C guy....

Something like:
`
            groups.sort {
                $0.name?.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
            entries.sort {
                $0.title()?.localizedCaseInsensitiveCompare($1.title()) == .orderedAscending
            }
`

**Comment by gabox7** [comment]
Ver.1,7.2 maybe works smothly ;) Ill test during this day. thanks in advance for fast fixed..

**Comment by dbush2017** [comment]
Same problem with 1.7.2 Crashes as soon as I open Group. Happens with two of five groups.

**Comment by dbush2017** [comment]
This is still an active issue for me. Anytime I attempt to open a group, the app crashes. Please advise when this will be addressed. It has been a month since I first reported the issue.

**Comment by tssmith7** [comment]
@dbush2017 is the database used with other programs on the PC/MAC or only MKP?  I would suggest two things to try:

  1. Open the database with KeePass 2.36 on a Windows PC and make some changes to the Group/Entries crashing and save. Then try opening in MKP. You could just add to the Notes or make a change, save, then change back and save again. The idea is to have KeePass write a database that MKP won’t crash with. 

2.  Take a copy of your database and delete all groups except a crashing group. Then slowly delete entries until the database doesn’t crash anymore. Try to eliminate as much as possible while still crashing. You will need to use another program on a PC or Mac. Ideally you could narrow it down to just a couple entries. If that is possible and there is few enough entries to remove your personal information and still have it crash then you could upload the database for debugging. 

Are you using KeePassX for Mac?  See @571

**Comment by dbush2017** [comment]
I am using KeePassX for Mac. It is version 0.4.3 on OSX 10.11.6. Everything works fine on the Max version. I haven't tried updating any entries. Just retrieving passwords since I can't reach them via my iPhone. iPhone is my master db. Mac is only for backup.
Donna

**Comment by tssmith7** [comment]
@dbush2017 see #571 as it appears there is a problem with OneDrive setting the file permissions incorrectly.  If you use OneDrive this could be the problem...

**Comment by dbush2017** [comment]
I don't use OneDrive

**Comment by dbush2017** [comment]
I don't use OneDrive.

**Comment by tssmith7** [comment]
@dbush2017, 

I some ideas/questions:

1. Create a new database with KeePassX 0.4.3 with a few entries and move it to your phone to see if MKP will open and edit the database.
2. How do you move your files from the phone to the Mac and back?
3. You can open some of the entries in your database (right?).  Can you open one of those entries and make changes to it?  Does MKP successfully save the changes?
4. Make some changes to your database with KeePassX on the Mac and move the resulting database to your phone.  Check if MKP will work.
